pandas/tests/tools/test_to_datetime.py

```diff
@@ -609,6 +609,119 @@ def test_to_datetime_dtarr(self, tz):
         result = to_datetime(arr)
         assert result is arr
 
+    @pytest.mark.parametrize(
+        "data, expected",
+        [
+            (Timestamp("2001-10-01 12:00:01.123456789"), Timestamp("2001-10-01 12:00:01.123456")),
+            (datetime(2002, 10, 1, 12, 00, 1, 123456), Timestamp("2002-10-01 12:00:01.123456")),
+            ("10/01/03 12:00:01.123456789", Timestamp("2003-10-01 12:00:01.123456789")),
+        ],
+    )
+    def test_to_datetime_preserves_resolution_when_possible(
+        self, data, expected
+    ):
+        # GH 49298
+        if not isinstance(data, str):
+            result = to_datetime([data])
+            tm.assert_equal(result, DatetimeIndex([data]))
+
+        result = to_datetime([data], format="%m/%d/%y %H:%M:%S.%f")
+        tm.assert_equal(result, DatetimeIndex([expected]))
+
+    @pytest.mark.parametrize(
+        "init_constructor, end_constructor",
+        [
+            (Index, DatetimeIndex),
+            (list, DatetimeIndex),
+            (np.array, DatetimeIndex),
+            (Series, Series),
+        ],
+    )
+    def test_to_datetime_arraylike_contains_pydatetime_and_timestamp(
+        self, init_constructor, end_constructor
+    ):
+        # GH 49298
+        # Timestamp/datetime have more resolution than str
+        case1 = [
+            Timestamp("2001-10-01 12:00:01.123456789"),
+            datetime(2001, 10, 2, 12, 30, 1, 123456),
+            "10/03/01",
+        ]
+        result = to_datetime(init_constructor(case1), format="%m/%d/%y")
+        expected_data = [
+            Timestamp("2001-10-01"),
+            Timestamp("2001-10-02"),
+            Timestamp("2001-10-03"),
+        ]
+        expected = end_constructor(expected_data)
+        tm.assert_equal(result, expected)
+
+        # Timestamp/datetime have the same resolution than str (nanosecond)
+        case2 = [
+            Timestamp("2001-10-01 12:00:01.123456789"),
+            datetime(2001, 10, 2, 12, 30, 1, 123456),
+            "10/03/01 13:00:01.123456789",
+        ]
+        result = to_datetime(init_constructor(case2), format="%m/%d/%y %H:%M:%S.%f")
+        expected_data = [
+            Timestamp("2001-10-01 12:00:01.123456"),
+            Timestamp("2001-10-02 12:30:01.123456"),
+            Timestamp("2001-10-03 13:00:01.123456789"),
```

Comment:
This might seem weird but it has an explanation for each case:

* `Timestamp.strftime` does not support nanoseconds: https://github.com/pandas-dev/pandas/issues/29461
* `datetime.datetime` also has the same problem: https://github.com/python/cpython/issues/59648